fplll/pruner.h

```diff
@@ -0,0 +1,568 @@
+          rv[i] * tabulated_ball_vol[i + 1] *
+          sqrt(pow_si(b[i / 2], 1 + i)) / pv[i];
+
+
+
+
+
+
+
+
+
+
+
+    total += tmp;
+  }
+  total /= symmetry_factor;  
+  //exit(1);
+  return total;
+}
+
+
+template<class FT>
+inline FT Pruner<FT>::svp_success_proba(/*i*/ evec &b){
+
+  evec b_minus_db;
+  FT dx = shell_ratio;
+
+  for (int i = 0; i < d; ++i) {
+    b_minus_db[i] = b[i] / (dx * dx);
+    if (b_minus_db[i] > 1)
+      b_minus_db[i] = 1;
+  }
+
+  FT vol = relative_volume(d, b);
+  FT dxn = pow_si(dx, 2 * d);
+  FT dvol =  dxn * relative_volume(d, b_minus_db) - vol;
+  return dvol / (dxn - 1.);
+
+}
+
+
+
+template<class FT>
+inline FT Pruner<FT>::cost_factor(/*i*/ evec &b){
+
+  FT success_proba = svp_success_proba(b);
+
+  if (success_proba >= target_success_proba)
+    return cost(b);
+
+  FT trials =  log(one - target_success_proba) / log(one - success_proba);
+  return cost(b) * trials + preproc_cost * (trials-1);
+}
+
+
+
+template<class FT>
+FT Pruner<FT>::cost_factor_derivative(/*i*/ evec &b, /*o*/ evec &res){
+  evec bpDb;
+  res[d - 1] = 0.;
+  for (int i = 0; i < d-1; ++i) {
+    bpDb = b;
+    bpDb[i] *= (one - epsilon);
+    enforce(bpDb, i);
+    FT X = cost_factor(bpDb);
+
+    bpDb = b;
+    bpDb[i] *= (one + epsilon);
+    enforce(bpDb, i);
+    FT Y = cost_factor(bpDb);
+
+
+
+    res[i] = (log(X) - log(Y)) / epsilon;
+  }
+
+}  
+template<class FT>
+int Pruner<FT>::improve(/*io*/ evec &b){
+
+  FT cf = cost_factor(b);
+  FT old_cf = cf;
+  evec newb;
+  evec gradient;
+  cost_factor_derivative(b, gradient);
+  FT norm = 0;
+
+  for (int i = 0; i < d; ++i)
+  {
+
+  }
+
+  // normalize the gradient
+  for (int i = 0; i < d; ++i) {
+    norm += gradient[i] * gradient[i];
+    newb[i] = b[i];
+  }
+
+  norm = sqrt(norm /  (one * (1. * d)) );
+  if (norm <= 0.)
+    return 0;
+  for (int i = 0; i < d; ++i)
+  {
+
+  }
+
+  for (int i = 0; i < d; ++i) {
+    gradient[i] /= norm;
+  }
+  FT new_cf;
+
+  FT step = min_step;
+  int i;
+
+  for (i = 0;; ++i) {
+    for (int i = 0; i < d; ++i) {
+      newb[i] = newb[i] + step * gradient[i];
+    }
+
+    enforce(newb);
+    new_cf = cost_factor(newb);
+
+
+
+    if (new_cf >= cf){
+      break;
+    }
+    b = newb;
+    cf = new_cf;
+    step *= step_factor;
+  }
+
+  if (cf > old_cf * min_cf_decrease){
+    return 0;
+  }
+  return i;
+}
+
+template<class FT>
+void Pruner<FT>::descent(/*io*/ evec &b){
+  while (improve(b)) { };
+}
+
+
+template<class FT>
+void Pruner<FT>::init_prunning_coeffs(evec &b) {
+  for (int i = 0; i < d; ++i) {
+    b[i] = .1 + ((1.*i) / d);
+  }
+  enforce(b);
+}
+
+
+
+
+
+/// Autoprune function, hiding the Pruner class
+
+
+template<class FT, class GSO_ZT, class GSO_FT> 
+void auto_prune(/*output*/ double* pr, double& success_proba,
+                /*inputs*/ double enumeration_radius, double preproc_cost,
+                double target_success_proba,
+                MatGSO<GSO_ZT, GSO_FT>& gso, int beginning = 0, int end = 0){
```

Comment:
I can surely find a simple way of re-adding this feature. Maybe for the next pull request ?